x-pack/build_chromium/build.py

```diff
@@ -1,55 +1,80 @@
-import subprocess, os, sys, platform, zipfile, hashlib, shutil
-from build_util import runcmd, mkdir, md5_file, script_dir, root_dir, configure_environment
+import os, subprocess, sys, platform, zipfile, hashlib, shutil
+from os import path
+from build_util import (
+  runcmd,
+  runcmdsilent,
+  mkdir,
+  md5_file,
+  configure_environment,
+)
 
 # This file builds Chromium headless on Windows, Mac, and Linux.
 
 # Verify that we have an argument, and if not print instructions
 if (len(sys.argv) < 2):
   print('Usage:')
-  print('python build.py {chromium_version}')
+  print('python build.py {chromium_version} [arch_name]')
   print('Example:')
   print('python build.py 68.0.3440.106')
   print('python build.py 4747cc23ae334a57a35ed3c8e6adcdbc8a50d479')
+  print('python build.py 4747cc23ae334a57a35ed3c8e6adcdbc8a50d479 arm64 # build for ARM architecture')
+  print
   sys.exit(1)
 
+src_path = path.abspath(path.join(os.curdir, 'chromium', 'src'))
+build_path = path.abspath(path.join(src_path, '..', '..'))
+build_chromium_path = path.abspath(path.dirname(__file__))
+argsgn_file = path.join(build_chromium_path, platform.system().lower(), 'args.gn')
+
 # The version of Chromium we wish to build. This can be any valid git
 # commit, tag, or branch, so: 68.0.3440.106 or
 # 4747cc23ae334a57a35ed3c8e6adcdbc8a50d479
 source_version = sys.argv[1]
+base_version = source_version[:7].strip('.')
 
 # Set to "arm" to build for ARM on Linux
 arch_name = sys.argv[2] if len(sys.argv) >= 3 else 'x64'
 
-print('Building Chromium ' + source_version + ' for ' + arch_name)
-
-# Set the environment variables required by the build tools
-print('Configuring the build environment')
-configure_environment()
-
-# Sync the codebase to the correct version, syncing master first
-# to ensure that we actually have all the versions we may refer to
-print('Syncing source code')
-
-os.chdir(os.path.join(root_dir, 'chromium/src'))
-
-runcmd('git checkout master')
-runcmd('git fetch origin')
-runcmd('gclient sync --with_branch_heads --with_tags --jobs 16')
-runcmd('git checkout ' + source_version)
-runcmd('gclient sync --with_branch_heads --with_tags --jobs 16')
-runcmd('gclient runhooks')
+if arch_name != 'x64' and arch_name != 'arm64':
+  raise Exception('Unexpected architecture: ' + arch_name)
+
+print('Building Chromium ' + source_version + ' for ' + arch_name + ' from ' + src_path)
+print('src path: ' + src_path)
+print('depot_tools path: ' + path.join(build_path, 'depot_tools'))
+print('build_chromium_path: ' + build_chromium_path)
+print('args.gn file: ' + argsgn_file)
+print
+
+# Sync the codebase to the correct version
+print('Setting local tracking branch')
+print(' > cd ' + src_path)
+os.chdir(src_path)
+
+checked_out = runcmdsilent('git checkout build-' + base_version)
+if checked_out != 0:
+  print('Syncing remote version')
+  runcmd('git fetch origin ' + source_version)
```

Comment:
This is a big improvement to initializing the working directory: only fetch the source version of the commit ID that we passed into the script